torch/distributed/tensor/parallel/_utils.py

```diff
@@ -46,15 +48,53 @@ def _validate_tp_mesh_dim(
-            )
+        # if tp_mesh_dim != 1:
+        #     raise RuntimeError(
+        #         f"Found TP device_mesh on the {tp_mesh_dim} dimension of its parent mesh.",
+        #         "Currently we only support intranode TP and TP needs to be the innermost dimension on its parent mesh.",
+        #     )
+
+def _check_tp_module_type(module, allowed_type):
+    """
+    Perform a check on `module` to see if it is of the allowed type.
+
+    Logically similar to running `isinstance(module, allowed_type)` but makes an additional check for module
+    of type InterpreterModule (from Pipeline Parallel tracing frontend) to find metadata on the node indicating it
+    was of the allowed type before tracing.
+
+    TODO(whc) this checker may be updated or not needed after changes in
+    https://github.com/pytorch/pytorch/issues/125245 land, which may restore the actual 'nn.Linear' object and restore
+    isinstance() functionality.
+    """
+
+    # foo = list(module.graph.nodes)[2]
+    # foo.meta
+    # [rank0]:(Pdb) [rank0]:(Pdb) [rank0]:{'stack_trace': '  File "/data/users/whc/torchtrain/torchtitan/models/llama/model.py", line 429, in forward\n    h = self.tok_embeddings(tokens)\n', 'nn_module_stack': {'L__self__': ('', 'torchtitan.models.llama.model.Transformer'), 'L__self___tok_embeddings': ('tok_embeddings', 'torch.nn.modules.sparse.Embedding')}, 'source_fn_stack': [('l__self___tok_embeddings', <class 'torch.nn.modules.sparse.Embedding'>)], 'original_aten': <OpOverload(op='aten.embedding', overload='default')>, 'from_node': [('h', 'L__self___tok_embeddings')], 'seq_nr': 437, 'torch_fn': ('embedding_1', 'function.embedding'), 'val': FakeTensor(..., device='meta', size=(4, 2048, 256)), 'tensor_meta': TensorMetadata(shape=torch.Size([4, 2048, 256]), dtype=torch.float32, requires_grad=False, stride=(524288, 256, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})}
+    # foo.meta['nn_module_stack']
+    # [rank0]:(Pdb) [rank0]:{'L__self__': ('', 'torchtitan.models.llama.model.Transformer'), 'L__self___tok_embeddings': ('tok_embeddings', 'torch.nn.modules.sparse.Embedding')}
+    # foo.meta.keys()
+    # [rank0]:(Pdb) [rank0]:dict_keys(['stack_trace', 'nn_module_stack', 'source_fn_stack', 'original_aten', 'from_node', 'seq_nr', 'torch_fn', 'val', 'tensor_meta'])
+
+    def has_allowed_metadata(module):
+        if isinstance(module, InterpreterModule):
+            found = False
+            for node in module.graph.nodes:
+                if hasattr(node, "meta"):
+                    meta = node.meta
+                    if 'nn_module_stack' not in meta:
+                        continue
+                    submod_name, submod_type = list(meta['nn_module_stack'].values())[-1]
+                    module, classname = submod_type.rsplit('.', 1)
+                    submod_class = getattr(importlib.import_module(module), classname)
```

Comment:
nit: if for each node, we do the import, how big would the overhead be? 

How about:

`if _check_tp_module_type(module, nn.Linear) or _check_tp_module_classname(module, "torch.nn.Linear")`

Reply:
Oh- the import was cleaned up already iirc, maybe forgot to push. 

I think it's easier to have just one function call for now but inside it I already return isinstance() or newcodepath() so the time spent on new code path should be avoided in non PP cases